enterprise/server/cmd/ci_runner/main.go

```diff
@@ -713,6 +714,26 @@ func parseFlags() error {
 			return err
 		}
 	}
+	if err := handleBackwardsCompatability(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// pushedRef should always be set, but old clients may not set it. Populate from
+// other flags if not set
```

Comment:
@bduffany  A downside of this implementation is that it adds a redundant "ref" field. It should always match one of "branch" or "commit", but if it doesn't, then there's ambiguity on which should be used. Another complication is that any or all of "ref", "branch", or "commit" can be set, which makes the code more complex.

Rather than adding "ref" flag to the ci_runner, I'm now leaning towards changing the API of ExecuteWorkflow. We could add a "ref_type" field, or replace "ref" with explicit "branch" or "commit" fields. Thoughts?

Reply:
> A downside of this implementation is that it adds a redundant "ref" field. It should always match one of "branch" or "commit", but if it doesn't, then there's ambiguity on which should be used.

I think in practice we should always have exactly one of { ref, branch+sha }, no?
* The external ExecuteWorkflow API takes a ref
* The internal ExecuteWorkflow API should (in practice) always be setting pushed_branch + commit_sha
* Webhook-triggered workflows should always have pushed_branch + commit_sha
* (IDK about remote bazel, but I guess we could just pass a ref?)

> Another complication is that any or all of "ref", "branch", or "commit" can be set, which makes the code more complex.

The CI runner code becomes more complex, but it buys a simpler API for users which IMO is worth the hidden complexity on our side of resolving the ref. Users are familiar with just doing `git checkout <ref>` where ref can be either a branch or a commit - and also the `actions/checkout` action on GH just needs a ref too.

In the case where the ref is a commit SHA, we do lose out on the branch name metadata with the current implementation, since we don't do a complete fetch of the repo. Though if people really want that metadata in their invocation, they could always set it explicitly via the separate metadata struct? Another idea is that we could make "complete fetch of the repo" the default behavior (I think GH actions does this?), and provide some more advanced options to allow fetching less data. I suspect that a complete fetch is fine (and probably more intuitive) except for very large repos, but these power users can use the more advanced controls to fetch less data at the cost of having to configure the branch metadata manually.

Reply:
External ExecuteWorkflow API is the only case where all 3 (ref, branch, and sha) could be set, if the user sets the metadata fields. In this case, we should checkout `ref`, because branch or sha are supposed to be metadata-only. This is why I was treating ref as the preferred choice

I think having the shallow clone is a better default, because it makes things much faster and customers probably shouldn't need the full git history

Reply:
> External ExecuteWorkflow API is the only case where all 3 (ref, branch, and sha) could be set, if the user sets the metadata fields. In this case, we should checkout ref, because branch or sha are supposed to be metadata-only. This is why I was treating ref as the preferred choice

wdyt about passing the metadata as separate args to the CI runner? For all the cases except for the ExecuteWorkflow API (which is less common to begin with), it seems like we want the branch+sha to be used for checkout and not just metadata. Whereas in the ExecuteWorkflow case, it's more ergonomic to provide a ref and then _maybe_ provide branch+sha just for metadata purposes?